src/validate.ts

```diff
@@ -26,18 +40,84 @@ function responseValidation(apiResponse: object, schema: object) {
 
 /**
  * Identifies the response type and validates the function based on the schemas.
- * @param apiResponse User defined api response
- * @param responseType User defined intent response
+ * @param req Request from one of the generate request functions.
+ * @param apiResponse User defined api response.
  * @return Errors from AJV validation, if any. Undefined otherwise.
  */
-export function validate(intentRequest: object, apiResponse: object) {
+export function validate(intentRequest: object, apiResponse: object, syncData?: object){
   const responseType = intentRequest['inputs'][0]['intent'];
   if (responseType === 'action.devices.SYNC') {
-    return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+    // validate traits with attributes schema
+    const syncErrors : object[] = [];
+    const syncDevices = apiResponse['payload']['devices'];
+    const syncDevicesLength = syncDevices.length;
+    for (let i = 0; i < syncDevicesLength; i++) {
+      const traits = syncDevices[i]['traits'];
+      const traitsLength = syncDevices[i]['traits'].length;
+      const attributes = syncDevices[i]['attributes'];
+      if (attributes === undefined) {
+        return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
+      }
+      for (let j = 0; j < traitsLength; j++) {
+        const trait = traits[j];
+        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
+          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
+          if (validateTraitRes) {
+            return syncErrors.push(...validateTraitRes);
+          }
+        }
+      }
+    }
+    return syncErrors.length ? syncErrors : undefined;
   } else if (responseType === 'action.devices.QUERY') {
+    // validate with states schema;
+    const queryErrors : object[] = [];
+    const devices = intentRequest['inputs'][0]['payload']['devices'];
+    const devicesLength = devices.length;
+    for (let i = 0; i < devicesLength; i++) {
+      const deviceIds = devices[i]['id'];
+      const states = apiResponse['payload']['devices'][deviceIds];
+      if (syncData != undefined){
+        const syncDevices = syncData['payload']['devices'];
+        const syncDevicesLength = syncDevices.length;
+        for (let j = 0; j < syncDevicesLength; j++) {
+             const trait = syncDevices[j]['traits'];
+             if (trait in TRAITS_COMMANDS_PAIR){
+               const validateQueryTraitStates = responseValidation(states, COMMAND_STATES_EXPECT[TRAITS_COMMANDS_PAIR[trait]]);
```

Comment:
Do we still need `TRAIT_COMMANDS_PAIR` if we're directly getting the traits and only the traits?

Reply:
For QUERY response, doesn't that only include the states schema?

Reply:
Can you associate a trait directly to expected states rather than using the indirection?

Reply:
We could maybe do a check to see if 'on off' is specified at the end instead and use that to identify what states to use?

Reply:
We can't rely on that in every case, such as [LightEffects](https://developers.google.com/assistant/smarthome/traits/lighteffects), so it may need to be a more explicit check, but I think you can use an object mapping trait to the expected states schema.

Reply:
Gotcha I pushed some changes using a different method, but maybe its easier to use the sync data for the execute response too so that 

const COMMAND_STATES_EXPECT = {
  'action.devices.traits.OnOff': ON_OFF_STATES_SCHEMA,
};